File: eng/docs/agent_skills/python_typing/examples/test_observability_client_example.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any
from unittest.mock import create_autospec

import pytest

# Adjust these imports to your real package path.
# They mirror the small unit shape discussed in the conversation.
from observability_client import (  # type: ignore[import-not-found]
    ObservabilityClient,
    ObservabilityClientConfig,
    ObservabilityClientPort,
    Result,
    SystemObservabilityPort,
)


class DummyCallContext:
    """Minimal test context for units that only require a typed ctx object."""

    pass


@dataclass(frozen=True, slots=True)
class Case:
    """Concrete logical test case rendered for pytest.

    This is intentionally small:
    - input axes: explicit/omitted visibility and data
    - dependency behavior axis: downstream success or failure
    - oracle: event shape + propagated Result
    """

    case_id: str
    title: str
    default_visibility: str
    ev: str
    severity: str
    visibility: str | None
    data: dict[str, Any] | None
    downstream_outcome: str
    expected_visibility: str
    expected_data: dict[str, Any]
    expected_ok: Any | None = None
    expected_err: Any | None = None


@pytest.fixture
def ctx() -> DummyCallContext:
    return DummyCallContext()


@pytest.fixture
def system_observability() -> SystemObservabilityPort:
    """Controlled collaborator boundary.

    We use a constrained mock so interface drift is easier to catch.
    """
    return create_autospec(SystemObservabilityPort, spec_set=True, instance=True)


@pytest.fixture
def make_sut(system_observability: SystemObservabilityPort):
    def _make(*, default_visibility: str = "internal") -> ObservabilityClient:
        return ObservabilityClient(
            observability=system_observability,
            config=ObservabilityClientConfig(default_visibility=default_visibility),
        )

    return _make


def arrange_downstream(case: Case, *, system_observability: SystemObservabilityPort) -> None:
    if case.downstream_outcome == "success":
        system_observability.emit.return_value = Result.success(case.expected_ok)
        return

    if case.downstream_outcome == "failure":
        system_observability.emit.return_value = Result(
            outcome="err",
            err=case.expected_err,
        )
        return

    raise ValueError(f"Unknown downstream_outcome: {case.downstream_outcome}")


SUCCESS_CASES = [
    Case(
        case_id="UTCASE_OBS_001",
        title="omitted visibility uses config default",
        default_visibility="internal",
        ev="TASK_STARTED",
        severity="INFO",
        visibility=None,
        data=None,
        downstream_outcome="success",
        expected_visibility="internal",
        expected_data={},
        expected_ok={"accepted": True},
    ),
    Case(
        case_id="UTCASE_OBS_002",
        title="explicit visibility overrides config default",
        default_visibility="internal",
        ev="TASK_STARTED",
        severity="INFO",
        visibility="both",
        data={"task_ref": "T_001"},
        downstream_outcome="success",
        expected_visibility="both",
        expected_data={"task_ref": "T_001"},
        expected_ok={"accepted": True},
    ),
    Case(
        case_id="UTCASE_OBS_003",
        title="omitted data becomes empty dict",
        default_visibility="both",
        ev="TASK_COMPLETED",
        severity="INFO",
        visibility=None,
        data=None,
        downstream_outcome="success",
        expected_visibility="both",
        expected_data={},
        expected_ok={"accepted": True},
    ),
]


FAILURE_CASES = [
    Case(
        case_id="UTCASE_OBS_004",
        title="downstream failure result is propagated",
        default_visibility="internal",
        ev="TASK_FAILED",
        severity="ERROR",
        visibility=None,
        data={"reason": "network_unavailable"},
        downstream_outcome="failure",
        expected_visibility="internal",
        expected_data={"reason": "network_unavailable"},
        expected_err={"code": "OBS_DOWN"},
    ),
]


@pytest.mark.parametrize("case", SUCCESS_CASES, ids=lambda c: c.case_id)
def test_emit_nominal_and_config_behavior(
    case: Case,
    ctx: DummyCallContext,
    make_sut,
    system_observability: SystemObservabilityPort,
) -> None:
    """Covers these logical families:
    - nominal_behavior
    - configuration_behavior
    - result_contract
    """
    sut = make_sut(default_visibility=case.default_visibility)
    arrange_downstream(case, system_observability=system_observability)

    result = sut.emit(
        ctx,
        ev=case.ev,
        severity=case.severity,
        visibility=case.visibility,
        data=case.data,
    )

    assert result.outcome == "ok"
    assert result.ok == case.expected_ok
    assert result.err is None

    system_observability.emit.assert_called_once()
    called_ctx, = system_observability.emit.call_args.args
    called_kwargs = system_observability.emit.call_args.kwargs

    assert called_ctx is ctx
    assert called_kwargs["event"] == {
        "ev": case.ev,
        "severity": case.severity,
        "visibility": case.expected_visibility,
        "data": case.expected_data,
    }


@pytest.mark.parametrize("case", FAILURE_CASES, ids=lambda c: c.case_id)
def test_emit_propagates_downstream_failure_result(
    case: Case,
    ctx: DummyCallContext,
    make_sut,
    system_observability: SystemObservabilityPort,
) -> None:
    """Covers these logical families:
    - dependency_behavior_classes
    - result_contract
    """
    sut = make_sut(default_visibility=case.default_visibility)
    arrange_downstream(case, system_observability=system_observability)

    result = sut.emit(
        ctx,
        ev=case.ev,
        severity=case.severity,
        visibility=case.visibility,
        data=case.data,
    )

    assert result.outcome == "err"
    assert result.ok is None
    assert result.err == case.expected_err

    system_observability.emit.assert_called_once_with(
        ctx,
        event={
            "ev": case.ev,
            "severity": case.severity,
            "visibility": case.expected_visibility,
            "data": case.expected_data,
        },
    )


def test_emit_never_mutates_input_data_dict(
    ctx: DummyCallContext,
    make_sut,
    system_observability: SystemObservabilityPort,
) -> None:
    """Example of a focused explicit test instead of parametrization.

    This is useful when the case is important and easier to read standalone.
    """
    payload = {"task_ref": "T_001"}
    system_observability.emit.return_value = Result.success({"accepted": True})
    sut = make_sut(default_visibility="internal")

    result = sut.emit(
        ctx,
        ev="TASK_STARTED",
        severity="INFO",
        data=payload,
    )

    assert result.outcome == "ok"
    assert payload == {"task_ref": "T_001"}

    system_observability.emit.assert_called_once_with(
        ctx,
        event={
            "ev": "TASK_STARTED",
            "severity": "INFO",
            "visibility": "internal",
            "data": {"task_ref": "T_001"},
        },
    )

File: eng/docs/agent_skills/csharp_typing/examples/TestObservabilityClientExample.cs
// ============================================================
// Unit under test: ObservabilityClient
// Namespace: Acme.Observability
// ============================================================
//
// Families covered:
//   - nominal_behavior        (required — public operation exists)
//   - result_contract          (required — success / failure shapes)
//   - configuration_behavior   (selected — default_visibility drives event shape)
//   - dependency_behavior_classes (selected — downstream failure propagated)
// Families skipped:
//   - input_admissibility      (skipped — no input validation at unit boundary)
//   - invariants               (skipped — no cross-path invariants beyond result_contract)
//   - outcome_error_mapping    (skipped — single error path, covered by dependency_behavior_classes)
//   - required_side_effects    (skipped — single emit call, covered by nominal_behavior assertions)
//   - forbidden_side_effects   (skipped — no conditional call skipping)
//   - initial_state_behavior   (skipped — stateless unit)
//   - ordering_protocol_behavior (skipped — single call, no ordering)

using FluentAssertions;
using NSubstitute;
using Xunit;

// Adjust these usings to your real namespace.
// They mirror the small unit shape discussed in the conversation.
using Acme.Observability;
using Acme.Observability.Ports;
using Acme.Observability.Types;

namespace Acme.Observability.Tests.Unit;

// ============================================================
// Case record
// ============================================================

internal sealed record Case
{
    public required string Label { get; init; }

    // -- inputs --
    public string DefaultVisibility { get; init; } = "internal";
    public string Ev { get; init; } = "TASK_STARTED";
    public string Severity { get; init; } = "INFO";
    public string? Visibility { get; init; } = null;
    public Dictionary<string, object>? Data { get; init; } = null;

    // -- collaborator configuration --
    public string DownstreamOutcome { get; init; } = "success";

    // -- expected result --
    public string ExpectedVisibility { get; init; } = "internal";
    public Dictionary<string, object> ExpectedData { get; init; } = new();
}

// ============================================================
// Test class
// ============================================================

public sealed class TestObservabilityClient
{
    private readonly ISystemObservabilityPort _systemObservability;

    public TestObservabilityClient()
    {
        _systemObservability = Substitute.For<ISystemObservabilityPort>();
    }

    // ---- SUT factory ----

    private ObservabilityClient MakeSut(string defaultVisibility = "internal")
    {
        var config = new ObservabilityClientConfig
        {
            DefaultVisibility = defaultVisibility,
        };

        return new ObservabilityClient(
            observability: _systemObservability,
            config: config);
    }

    // ============================================================
    // Arrangement helpers
    // ============================================================

    private void ArrangeCase(Case c)
    {
        if (c.DownstreamOutcome == "failure")
        {
            _systemObservability
                .When(x => x.Emit(Arg.Any<ObservabilityEvent>()))
                .Do(_ => throw new InvalidOperationException("observability unavailable"));
        }
    }

    private void AssertEmitCalledWith(Case c)
    {
        _systemObservability.Received(1).Emit(
            Arg.Is<ObservabilityEvent>(e =>
                e.Ev == c.Ev &&
                e.Severity == c.Severity &&
                e.Visibility == c.ExpectedVisibility));
    }

    // ============================================================
    // Case tables
    // ============================================================

    public static IEnumerable<object[]> SuccessCases =>
    [
        [new Case
        {
            Label = "omitted_visibility_uses_config_default",
            DefaultVisibility = "internal",
            Ev = "TASK_STARTED",
            Severity = "INFO",
            Visibility = null,
            Data = null,
            DownstreamOutcome = "success",
            ExpectedVisibility = "internal",
            ExpectedData = new(),
        }],
        [new Case
        {
            Label = "explicit_visibility_overrides_config_default",
            DefaultVisibility = "internal",
            Ev = "TASK_STARTED",
            Severity = "INFO",
            Visibility = "both",
            Data = new() { { "task_ref", "T_001" } },
            DownstreamOutcome = "success",
            ExpectedVisibility = "both",
            ExpectedData = new() { { "task_ref", "T_001" } },
        }],
        [new Case
        {
            Label = "omitted_data_becomes_empty_dictionary",
            DefaultVisibility = "both",
            Ev = "TASK_COMPLETED",
            Severity = "INFO",
            Visibility = null,
            Data = null,
            DownstreamOutcome = "success",
            ExpectedVisibility = "both",
            ExpectedData = new(),
        }],
    ];

    public static IEnumerable<object[]> FailureCases =>
    [
        [new Case
        {
            Label = "downstream_failure_result_is_propagated",
            DefaultVisibility = "internal",
            Ev = "TASK_FAILED",
            Severity = "ERROR",
            Visibility = null,
            Data = new() { { "reason", "network_unavailable" } },
            DownstreamOutcome = "failure",
            ExpectedVisibility = "internal",
            ExpectedData = new() { { "reason", "network_unavailable" } },
        }],
    ];

    // ============================================================
    // Test methods
    // ============================================================

    [Theory]
    [MemberData(nameof(SuccessCases))]
    public void EmitNominalAndConfigBehavior(Case c)
    {
        // Arrange
        ArrangeCase(c);
        var sut = MakeSut(defaultVisibility: c.DefaultVisibility);

        // Act
        sut.Emit(
            ev: c.Ev,
            severity: c.Severity,
            visibility: c.Visibility,
            data: c.Data);

        // Assert
        AssertEmitCalledWith(c);
    }

    [Theory]
    [MemberData(nameof(FailureCases))]
    public void EmitPropagatesDownstreamFailureResult(Case c)
    {
        // Arrange
        ArrangeCase(c);
        var sut = MakeSut(defaultVisibility: c.DefaultVisibility);

        // Act
        var act = () => sut.Emit(
            ev: c.Ev,
            severity: c.Severity,
            visibility: c.Visibility,
            data: c.Data);

        // Assert
        act.Should().Throw<InvalidOperationException>();
        AssertEmitCalledWith(c);
    }

    [Fact]
    public void EmitNeverMutatesInputDataDictionary()
    {
        // Arrange — stand-alone invariant test
        var payload = new Dictionary<string, object> { { "task_ref", "T_001" } };
        var sut = MakeSut(defaultVisibility: "internal");

        // Act
        sut.Emit(
            ev: "TASK_STARTED",
            severity: "INFO",
            data: payload);

        // Assert
        payload.Should().BeEquivalentTo(
            new Dictionary<string, object> { { "task_ref", "T_001" } },
            "input data must not be mutated");
    }
}
